### DISCUSSION TITLE
Menu improvements master topic

### DISCUSSION THREAD

**Discussion by xezon** [TheSuperHackers/GeneralsGamePatch#686] [General]
We need to get an overview of what we can and want to change in the game menus. There is a lot of room for improvements, because the original game menus are not well polished.

Potential action items:
- Fix ui element position errors
- Reorganize ui element positions
- Remove obsolete ui elements
- Add hidden ui elements (for example AA setting in Options Menu, player list in LAN Lobby)
- Add text elements to help explain functionality, such as "Alternative Mouse Setup" in Options
- Improve textures (for example Zero Hour logo looks bad, score screen does not show accurate faction icons)
- Improve functionality (for example increase list box sizes and limits)

I suggest to take a look at all menu windows and note down all high level action items that become apparent.
We can also look at various other Mods to get inspiration for potential improvements.
We can also check if there is a Mod with menu changes that is suitable to use as a starting point.

**Comment by xezon** [comment]
Main Menu or Options menu should print new game version label confirming the installation of Patch104p. Ideally original version number label (1.00 .. 1.04) is preserved.

**Comment by MTKing4** [comment]
One thing to keep in mind which is the Main menu options' positions, we don't want to brake Gameranger's auto clicker

**Comment by xezon** [comment]
Stubbjax:
> I know a few people who passed on the game due to the fact that left-click was the action button (by default), which was confusing compared to the more standardised right-click in most other RTS games at the time. They didn't realise it could be changed until it was too late.

**Comment by xezon** [comment]
We can make a pass on all WND files and correct WINDOWTYPE erros. See:
https://github.com/xezon/GeneralsControlBar/commit/9abc7697c7ec2ade61e96513e040813c6a3234fb

We can also fix indentations in WND files. See:
https://github.com/xezon/GeneralsControlBar/commit/9d5edaf63978e926a79a91f9cd1fc72021742fad

**Comment by xezon** [comment]
We should add new HeaderTemplates to all SCROLLLISTBOX elements, so that font upscales do not break them.

**Comment by xezon** [comment]
A very common issue report is the "I cannot connect to online" when the wrong IP is selected in the Game Options menu. We could mention this common source of error in the Connection Failed popup and/or in the Connection screen. This way user may be better guided to the cause of the issue.

**Comment by Madrageeeeee** [comment]
Actually unsure whether it already exists butnis there a hotkey to make the general promotion screen pop-up? 

That would be nice (if im right that it doesnt currently exist)

**Reply by xezon** [reply]
As far as I am aware this is not possible.

**Comment by xezon** [comment]
# A brief Zero Hour menu color analysis

## Splash Screen

### Generals
Generals invites us with a dark and muted splash screen that makes the eye wander right to the center logo and the tanks beneath. Just the EA Logo does not fit in so well despite being integrated into a border. It appears to bright and white in comparison the the darker tone of the image. Still is a simple yet clean art piece that is good to look at. The colors make sense.

![Install_Final](https://user-images.githubusercontent.com/4720891/180802043-c9419418-0066-4e43-9379-ba45646eb1c4.jpg)

### Zero Hour
Zero Hour gives us a quite bright and entirely vibrant blue tinted splash screen resembling the world map. It sets the tone of what we can get to expect from the menus art. The artwork convinces less with its lack of contrast between the dark and the light and also is arguably less interesting to look at than the former splash.

![Install_Final2](https://user-images.githubusercontent.com/4720891/180802067-20d0ba25-bbda-44d2-a0da-b1381694e877.jpg)

## Main Menu

### Generals
The menu is Orange & Brown themed. Colors look consistent and mute. They match what we got to expect from the Splash screen. The button font color is irritating when not hovered, because it is much darker than its border. The logo looks presentable.

![shot_20220725_162311_1](https://user-images.githubusercontent.com/4720891/180805443-16705d73-b8a5-43aa-b571-4e0ee52803c3.jpg)

### Zero Hour
The menu is Blue & White themed. There is a noticeable color mismatch between the outer frame and the inner buttons. The Game logo matches colors with the buttons, but is visually inferior to the one of Generals. All text is white. Albeit the White text color is rarely a bad choice due its simplicity and visibility, it does not convey the same muted feel of the original Generals Theme. It is more clinical, nothing to be overly excited about. Hovering buttons will change the White text to Lime Green, which is mismatching the vibrant Blue tone. Albeit the eye gets used to it, it does not fit. The Generals Theme makes a better job of color consistency here.

![shot_20220725_163707_4](https://user-images.githubusercontent.com/4720891/180805476-85c76dfb-950e-4606-a634-f8ae38c34ece.jpg)

## More menu

### Generals
Flaws become more apparent when diving into menus. Why do buttons have darker texts in them than their borders? Why are some texts White, when others are Orange or Brown? The Radio Button next to "Alternate Mouse Setup" has a much more logical color order than the text buttons at the bottom. The sliders are too dark, mismatching with the brighter inner of the Radio Button.

![shot_20220725_161019_1](https://user-images.githubusercontent.com/4720891/180806941-67cfcc10-f86c-46de-b521-2dd0883c3a08.jpg)

### Zero Hour
The menu does not suffer from mismatching text colors. All text is pure white. The sliders appear to have more luminance, which matches its surroundings more closely. The background of the sliders still is dark brown, which is a remainder of the Generals Theme and does not fit into the Blue theme. The menu borders have become a lot darker and also present a mismatching Blue tone, similar to the global main ruler backdrop at the very sides. With this, the menu borders are not as well defined as they are in Generals. Buttons and Dropdownboxes have slightly mismatching blue tones. The arrow color inside the Dropdownbox is also mismatching. Generals Theme colors look much more consistent and pleasing in comparison.

![shot_20220725_161057_1](https://user-images.githubusercontent.com/4720891/180807015-8b22a255-70b2-4c6d-a4d8-57b2cd364bda.jpg)

## Ingame

### Zero Hour
The UI design carries over into the core game loop by having its blue theme on fundamental places, such as the UI Tooltip. The blue tint is visually in stark contrast to the faction specific Control Bars.

![shot_20220725_161827_3](https://user-images.githubusercontent.com/4720891/180817801-2817b308-a920-491c-a235-fd8d26441e76.jpg)

![shot_20220725_161851_5](https://user-images.githubusercontent.com/4720891/180818187-c0b5f497-7401-4c7d-9a63-1dcf289fbc27.jpg)

The chat box presents itself partly green tinted, which does not just mismatch the Control Bars, but also the blue theme.

![shot_20220725_162505_2](https://user-images.githubusercontent.com/4720891/180817979-68451642-43b3-4208-ba95-7ef4bf41facb.jpg)

## Details

### Zero Hour
Here are the dominant Blue color tones of UI elements. They typically sit very close together, with no gradients between them. These colors are unable to provide a pleasant visual presentation when used in combination with black backdrops and white texts.

![menucolors](https://user-images.githubusercontent.com/4720891/180822041-fbd691dc-cb32-4533-8c1f-d3931672b563.png)

**Reply by xezon** [reply]
To shape better color palette, a starting point is to take that strong blue and that bright cyan and take a mid point of these as the base color.

https://meyerweb.com/eric/tools/color-blend/#00E3FF:2534FF:8:rgbd

Here, rgb(21,130,255) could be a good pick, which resembles a blend between these original dominant menu colors. Next we can build a new primary color palette around our dominant color.

https://colors.muz.li/color/1582FF

"Mono" palette looks good for keeping it pure blue. "Complementary" color palette looks good if we want to introduce a bit more color into the mix. We can also generate "random" palette from dominant color.

http://colormind.io/

We dial #1582ff in fist color slot and lock the color, then hit Generate button. It will generate palette.
![palette1](https://user-images.githubusercontent.com/4720891/180965962-3d880576-5bc4-4b76-9fa5-29fa69e7037b.png)

A less vibrant color palette.
![palette2](https://user-images.githubusercontent.com/4720891/180966037-4885fb55-9d39-4db4-ada7-38fdf4ea973d.png)

Blue To Orange.
![palette3](https://user-images.githubusercontent.com/4720891/180966248-48413343-ed01-4b17-bdf6-1dbea55529f3.png)

Blue to Brown. It is a bit of a mix between ZH and Generals colors. We can probably optimize colors for this Generals Gold color. Its code is #ffbe00
![palette4](https://user-images.githubusercontent.com/4720891/180966367-373750c8-1c59-41c9-baed-82dde143bb3d.png)

Here is adjusted to Generals Gold.
![palette5](https://user-images.githubusercontent.com/4720891/180975857-ef84300e-f5a5-4811-b3a7-1ef343c2d178.png)

**Comment by samosaphile** [comment]
Possibly investigate the source of high power usage when staring at menu or lobby in zh. For some reason, fan usage/noise runs abnormally high on laptops, when they shouldn't. Even running a newer game like RA3 (in menu/lobby), runs the fans less. Wish it could be fanless.

**Comment by ExiLe-ZH** [comment]


**Reply by xezon** [reply]
Replay list bug is here
* #658

**Reply by ExiLe-ZH** [reply]
Another improvement would be adding preview information to the replay list/menu, like a map thumbnail, player names and teams. Basically similar to the actual multiplayer lobby but greyed out.

**Comment by DevGeniusCode** [comment]
Hi everyone,  

I wrote a script to scan the `OptionsMenu.wnd` file and extract all the colors used within it. Here's the output:  
![colors_palette](https://github.com/user-attachments/assets/2311d3be-de22-4888-8929-0a39a5b5a282)


After reviewing previous discussions on UI colors, I had an idea to use the color palette from the equivalent file in Generals (not ZH) as a baseline. My plan is to compare the palettes and establish correspondences between them.  
For example:  
- If a text color in Generals is defined as X, its equivalent in ZH could be determined based on the existing blue-white palette and through calculated adjustments.  
- For instance, the orange text in Generals currently maps to white in ZH, but it might work better as blue, provided we verify screen contrast and readability.  

### Progress and Findings:  
1. **Color List Simplification**  
   - I started by removing colors with no visual impact, such as:  
     - Colors with very close values (e.g., `0 0 2 255` is visually identical to `0 0 0 255`).
     - Colors like `168 255 12 0`, which have zero alpha and don't display on the screen,  full transparency.  
   - I also replaced all instances of `255 255 255 0` with `0 0 0 0`, which resulted in a **144,864-bit reduction in file size** for this file.  

2. **Missing Colors and Image-Based Elements**  
![image](https://github.com/user-attachments/assets/ba7e3fd3-4f04-4dca-8e91-0076e49cb7c6)

   - I noticed that the **brown color of the slider** and the **blue color of the slider blocks** are not defined in the file (The brown color does not appear at all in the color list).   
     - The blue slider blocks appear unrelated to the blue colors in the list.  
     - This suggests that, similar to button frames, `ComboBox` frames, and text box frames, the slider might be composed of small image elements (e.g., brown and blue images).  

3. **Untraceable Colors in `ComboBox`**  
   - Another interesting case is the **white text inside the `ComboBox`**, which remains unaffected even after removing all white colors from the file.  
   - This indicates that the white text is not defined in the color list and might be influenced by another source, similar to the brown color of the slider.  

4. **Final Cleaned-Up List**  

   After filtering out non-displayed and unrelated colors, here's the resulting list:  

   ![colors_palette](https://github.com/user-attachments/assets/bc537c21-a944-4429-b083-5e0a042a63cf)

   (Note: The different shades of black represent varying levels of transparency.)  

Would appreciate any thoughts or insights, especially regarding where additional default colors might be defined or how best to approach the palette mapping.

**Reply by xezon** [reply]
Many of the ui elements are inside Data\English\Art\Textures\SCSmShellUserInterface512_001.tga

> Another interesting case is the white text inside the ComboBox, which remains unaffected even after removing all white colors from the file.

Perhaps font color is taken from somewhere in Data\INI\InGameUI.ini

We can also dig in Thyme to see where ComboBox color might come from. I think Combo Box is implemented.

In the worst case, the white color is hardcoded in the executable.

> I also replaced all instances of 255 255 255 0 with 0 0 0 0, which resulted in a 144,864-bit reduction in file size for this file.

Are we convinced that these changes would not bite us later somehow? Are they for research purposes only or do you like this to be merged into main branch?

> This suggests that, similar to button frames, ComboBox frames, and text box frames, the slider might be composed of small image elements (e.g., brown and blue images).

It is definitely inside SCSmShellUserInterface512_001.tga

**Reply by DevGeniusCode** [reply]
> > I also replaced all instances of 255 255 255 0 with 0 0 0 0, which resulted in a 144,864-bit reduction in file size for this file.
>
> Are we convinced that these changes would not bite us later somehow? Are they for research purposes only or do you like this to be merged into main branch?

For research purposes only, I believe there is no risk, as the colors are fully transparent and not displayed at all. 
Of course, similar to any color changes, if we decide to reset the color codes, we will do so for all files in a dedicated branch, allowing us to revert and rallback in case of any issues.

**Comment by DevGeniusCode** [comment]
Here’s my revised proposal regarding the colors:

I based my approach on how the colors were applied in *Generals (1)*. In the regular state, the text color was similar to the background color of the highlighted state, but slightly lighter. For the highlighted button, the text color matched the middle shade of the button’s border.

Similarly, I’ve defined the colors for *ZH (2)*, though I haven't yet finalized the text color for the regular state. In my view, Option 4 appears to be the most suitable, as it aligns with the color of the CheckBox. Please see the example below.

But it can be argued that the regular state in option 4 is still too dark and will be difficult to read

![buttons](https://github.com/user-attachments/assets/0d8fbb44-b5fa-4cb6-9dba-ddbdcf5fe132)

**Reply by xezon** [reply]
Here are the Luminance values of the fonts:

Generals Original (1)
Highlight: 82
Default: 54

Zero Hour Original (2)
Highlight: 93
Default: 100

Zero Hour Edited (4)
Highlight: 76
Default: 43

It would be good to at least hit luminance levels of Generals, ideally exceed them because the original Zero Hour fonts are much brighter (white). The problem I see here is that blue color comes with the disadvantage of low luminance, which is a major disadvantage over greens and reds. I do not have a good idea right now. I think it is worthwhile to experiment more and try to find a color composition that looks nice.

**Comment by DevGeniusCode** [comment]
> * Remove obsolete ui elements
> * Add hidden ui elements (for example AA setting in Options Menu, player list in LAN Lobby)

List of removed/non-working features from the site [tcrf.net](https://tcrf.net/Command_%26_Conquer:_Generals/Unused_Content_in_the_Original#Unused_GUI_Elements)

# Replay Control
- Related https://github.com/TheSuperHackers/GeneralsGameCode/issues/285

![GensReplayControl.png](https://tcrf.net/images/1/15/GensReplayControl.png)

Replay control was meant as a way to control replays. You could make the replay play faster or slower, pause the replay, or enable frame advance mode. It was removed from the game though, leaving you with no way to control the replay, but in the Expansion Pack a fast forward mode was introduced as a makeshift replacement.

# Main Menu

![The Main Menu, viewed in a game editor.](https://tcrf.net/images/thumb/2/29/GensMainMenu.png/300px-GensMainMenu.png)

The Main Menu also offers some interesting hidden leftovers.

Apparently, it was planned that map packs would be regularly released which you could then download from the game. A hidden button "Download Map Pack" exists in the Main Menu and would presumably be used for this purpose. There's also an entry in the Windows registry called `MapPackVersion` which is presumably associated with it. It is always set to 10000, though.  
Similarly, updates were also planned to be downloaded from the Main Menu, as hinted by another hidden button "Get Update". But since updates cannot be installed anyway while the game is running, it was moved to the autorun menu.  
A clock was also intended to be visible in the Main Menu, as a small gimmick.

# Options Menu
- Related https://github.com/TheSuperHackers/GeneralsGamePatch/issues/2569

The options menu is surprisingly full of unused stuff.

- A hidden drop-down box would have allowed you to change the anti-aliasing. This is not possible in the final version.
- Apparently, the way you moved the camera was saved in replays and then reflected when the replay is run. This would have been useful for example for analyzing the player's navigation, but it has been removed from the final game.
- There is an option to disable the language filter! Anyone who has ever played an online game knows how annoying these are. This was presumably only intended for the game developers, and normal users probably were never intended to be able to switch it off.
- There is a button that reads "Keyboard Options". You may have been able to change the keyboard layout in this special menu. This would also explain the large amount of unused text relating to keyboard commands in the game, and also why the `CommandMap.ini` file, which controls the keyboard commands, is not saved in the game's data files, but instead extracted in one of the game's subfolders to be easily editable. The reason why it was removed may be to prevent keyboard commands from conflicting with the in-game button hotkeys, which cannot be edited this way.
- One hidden option in the advanced options menu disables the frame limiter, which normally limits the game speed to 30 FPS (or whatever you have it set to in the battle's options). The description says it is mainly intended for benchmarks and not for actual gameplay, because it does not just change the frame rate but also speeds up/slows down the game.
- There is a hidden text string that reads "Mouse Type". It's possible it was intended to allow you to use the fourth and fifth mouse buttons in the game (if you have them), but ultimately its purpose is uncertain.
- There are two hidden options "Draw scroll anchor" and "Move scroll anchor". The effects of these options are unknown.
